In [11]:
import pandas as pd 
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OrdinalEncoder
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import LabelEncoder


In [12]:
df=pd.read_csv('df_clean.csv')

In [13]:
df.PUNT_GLOBAL.median()

0.5

In [14]:
prueba=df.copy()

In [15]:
# features 
X=prueba[['ESTU_GENERO', 'EDAD', 'FAMI_ESTRATOVIVIENDA', 'FAMI_PERSONASHOGAR',
       'FAMI_EDUCACIONMADRE', 'FAMI_TIENEINTERNET', 'FAMI_TIENECOMPUTADOR',
       'ESTU_DEDICACIONLECTURADIARIA', 'COLE_MCPIO_UBICACION',
       'COLE_DEPTO_UBICACION']]

#target 

y=prueba.PUNT_GLOBAL

In [16]:
X.dtypes

ESTU_GENERO                     object
EDAD                             int64
FAMI_ESTRATOVIVIENDA            object
FAMI_PERSONASHOGAR              object
FAMI_EDUCACIONMADRE             object
FAMI_TIENEINTERNET              object
FAMI_TIENECOMPUTADOR            object
ESTU_DEDICACIONLECTURADIARIA    object
COLE_MCPIO_UBICACION            object
COLE_DEPTO_UBICACION            object
dtype: object

In [17]:
X_train,X_valid,y_train,y_valid=train_test_split(X,y,random_state=42,train_size=0.7,test_size=0.3)

### ENCODING FEATURES OF X_TRAIN

In [18]:
cat_ord=X_train[['EDAD','FAMI_ESTRATOVIVIENDA','FAMI_PERSONASHOGAR']]
cat_nom=X_train[['ESTU_GENERO','FAMI_TIENEINTERNET','FAMI_TIENECOMPUTADOR']]
cat_lab_muni=X_train[['COLE_MCPIO_UBICACION']]
cat_lab_dptos=X_train[['COLE_DEPTO_UBICACION']]
cat_per=X_train[['FAMI_EDUCACIONMADRE','ESTU_DEDICACIONLECTURADIARIA']]



#### Enconding nominal categories

In [19]:
onehot_encoder = OneHotEncoder(sparse=False)
onehot_encoded = onehot_encoder.fit_transform(cat_nom)
df_onehot = pd.DataFrame(onehot_encoded, columns=onehot_encoder.get_feature_names_out(['ESTU_GENERO','FAMI_TIENEINTERNET','FAMI_TIENECOMPUTADOR']))


C:\Users\pedro\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\preprocessing\_encoders.py:868: FutureWarning:

`sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.



#### Enconding Ordianal categories


In [20]:
ordinal_encoder = OrdinalEncoder()
ordinal_encoded = ordinal_encoder.fit_transform(cat_ord)
df_ordinal = pd.DataFrame(ordinal_encoded, columns=['EDAD','FAMI_ESTRATOVIVIENDA','FAMI_PERSONASHOGAR'])


#### Encoding labeled categories Municipios and departementos

In [21]:
labeled_encoder=LabelEncoder()
labeled_encoded=labeled_encoder.fit_transform(cat_lab_muni)
df_labeled_muni = pd.DataFrame(labeled_encoded, columns=['COLE_MCPIO_UBICACION'])

C:\Users\pedro\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\preprocessing\_label.py:116: DataConversionWarning:

A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().



In [22]:
labeled_encod=LabelEncoder()
labeled_encoded=labeled_encod.fit_transform(cat_lab_dptos)
df_labeled_dptos = pd.DataFrame(labeled_encoded, columns=['COLE_DEPTO_UBICACION'])

C:\Users\pedro\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\preprocessing\_label.py:116: DataConversionWarning:

A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().



#### Encoding 'FAMI_EDUCACIONMADRE','ESTU_DEDICACIONLECTURADIARIA'

In [23]:
orden_categorias = [
    'Ninguno',
    'Primaria incompleta',
    'Primaria completa',
    'Secundaria (Bachillerato) incompleta',
    'Secundaria (Bachillerato) completa',
    'Técnica o tecnológica incompleta',
    'Técnica o tecnológica completa',
    'Educación profesional incompleta',
    'Educación profesional completa',
    'Postgrado'
]

# Crear una instancia del codificador ordinal
OE = OrdinalEncoder(categories=[orden_categorias])

cd=OE.fit_transform(X_train[['FAMI_EDUCACIONMADRE']])
df_fami_educacion_madre = pd.DataFrame(cd, columns=['FAMI_EDUCACIONMADRE'])


In [24]:
orden_cat=['No leo por entretenimiento',
           '30 minutos o menos',
           'Entre 30 y 60 minutos',
           'Entre 1 y 2 horas',
           'Más de 2 horas']

OE_e=OrdinalEncoder(categories=[orden_cat])
pe=OE_e.fit_transform(X_train[['ESTU_DEDICACIONLECTURADIARIA']])
df_lect_dia=pd.DataFrame(pe, columns=['ESTU_DEDICACIONLECTURADIARIA'])

#### LETS JOIN DF'S

In [25]:
X_train_final = pd.concat([df_onehot, df_ordinal, df_labeled_muni,df_labeled_dptos,df_fami_educacion_madre,df_lect_dia], axis=1)


### ENCODING FEATURES OF X_VALID

In [26]:
cat_ord=X_valid[['EDAD','FAMI_ESTRATOVIVIENDA','FAMI_PERSONASHOGAR']]
cat_nom=X_valid[['ESTU_GENERO','FAMI_TIENEINTERNET','FAMI_TIENECOMPUTADOR']]
cat_lab_muni=X_valid[['COLE_MCPIO_UBICACION']]
cat_lab_dptos=X_valid[['COLE_DEPTO_UBICACION']]
cat_per=X_valid[['FAMI_EDUCACIONMADRE','ESTU_DEDICACIONLECTURADIARIA']]



# %% [markdown]
# #### Enconding nominal categories

# %%
onehot_encoder = OneHotEncoder(sparse=False)
onehot_encoded = onehot_encoder.fit_transform(cat_nom)
df_onehot = pd.DataFrame(onehot_encoded, columns=onehot_encoder.get_feature_names_out(['ESTU_GENERO','FAMI_TIENEINTERNET','FAMI_TIENECOMPUTADOR']))


# %% [markdown]
# #### Enconding Ordianal categories
# 

# %%
ordinal_encoder = OrdinalEncoder()
ordinal_encoded = ordinal_encoder.fit_transform(cat_ord)
df_ordinal = pd.DataFrame(ordinal_encoded, columns=['EDAD','FAMI_ESTRATOVIVIENDA','FAMI_PERSONASHOGAR'])


# %% [markdown]
# #### Encoding labeled categories Municipios and departementos

# %%
labeled_encoder=LabelEncoder()
labeled_encoded=labeled_encoder.fit_transform(cat_lab_muni)
df_labeled_muni = pd.DataFrame(labeled_encoded, columns=['COLE_MCPIO_UBICACION'])

# %%
labeled_encod=LabelEncoder()
labeled_encoded=labeled_encod.fit_transform(cat_lab_dptos)
df_labeled_dptos = pd.DataFrame(labeled_encoded, columns=['COLE_DEPTO_UBICACION'])

# %% [markdown]
# #### Encoding 'FAMI_EDUCACIONMADRE','ESTU_DEDICACIONLECTURADIARIA'

# %%
orden_categorias = [
    'Ninguno',
    'Primaria incompleta',
    'Primaria completa',
    'Secundaria (Bachillerato) incompleta',
    'Secundaria (Bachillerato) completa',
    'Técnica o tecnológica incompleta',
    'Técnica o tecnológica completa',
    'Educación profesional incompleta',
    'Educación profesional completa',
    'Postgrado'
]

# Crear una instancia del codificador ordinal
OE = OrdinalEncoder(categories=[orden_categorias])

cd=OE.fit_transform(X_valid[['FAMI_EDUCACIONMADRE']])
df_fami_educacion_madre = pd.DataFrame(cd, columns=['FAMI_EDUCACIONMADRE'])


# %%
orden_cat=['No leo por entretenimiento',
           '30 minutos o menos',
           'Entre 30 y 60 minutos',
           'Entre 1 y 2 horas',
           'Más de 2 horas']

OE_e=OrdinalEncoder(categories=[orden_cat])
pe=OE_e.fit_transform(X_valid[['ESTU_DEDICACIONLECTURADIARIA']])
df_lect_dia=pd.DataFrame(pe, columns=['ESTU_DEDICACIONLECTURADIARIA'])

# %% [markdown]
# #### let's join the df's

# %%
X_valid_final = pd.concat([df_onehot, df_ordinal, df_labeled_muni,df_labeled_dptos,df_fami_educacion_madre,df_lect_dia], axis=1)

C:\Users\pedro\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\preprocessing\_encoders.py:868: FutureWarning:

`sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.

C:\Users\pedro\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\preprocessing\_label.py:116: DataConversionWarning:

A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().

C:\Users\pedro\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\preprocessing\_label.py:116: DataConversionWarning:

A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for ex

In [27]:
X_valid_final

ESTU_GENERO_F  ESTU_GENERO_M  FAMI_TIENEINTERNET_No  \
0                 0.0            1.0                    1.0   
1                 0.0            1.0                    0.0   
2                 0.0            1.0                    0.0   
3                 0.0            1.0                    1.0   
4                 1.0            0.0                    1.0   
...               ...            ...                    ...   
130093            1.0            0.0                    0.0   
130094            1.0            0.0                    1.0   
130095            0.0            1.0                    0.0   
130096            1.0            0.0                    1.0   
130097            1.0            0.0                    0.0   

        FAMI_TIENEINTERNET_Si  FAMI_TIENECOMPUTADOR_No  \
0                         0.0                      0.0   
1                         1.0                      0.0   
2                         1.0                      0.0   
3                         0.0                      1.0   
4                         0.0                      0.0   
...                       ...                      ...   
130093                    1.0                      0.0   
130094                    0.0                      1.0   
130095                    1.0                      0.0   
130096                    0.0                      1.0   
130097                    1.0                      0.0   

        FAMI_TIENECOMPUTADOR_Si  EDAD  FAMI_ESTRATOVIVIENDA  \
0                           1.0   2.0                   0.0   
1                           1.0   1.0                   1.0   
2                           1.0   2.0                   2.0   
3                           0.0   1.0                   1.0   
4                           1.0   1.0                   0.0   
...                         ...   ...                   ...   
130093                      1.0   2.0                   1.0   
130094                      0.0   2.0                   1.0   
130095                      1.0   0.0                   1.0   
130096                      0.0   2.0                   4.0   
130097                      1.0   1.0                   1.0   

        FAMI_PERSONASHOGAR  COLE_MCPIO_UBICACION  COLE_DEPTO_UBICACION  \
0                      1.0                   229                     6   
1                      2.0                   815                    19   
2                      1.0                   510                     1   
3                      2.0                    99                     4   
4                      0.0                   510                     1   
...                    ...                   ...                   ...   
130093                 2.0                   106                    27   
130094                 2.0                   128                    30   
130095                 1.0                    99                     4   
130096                 1.0                   208                    12   
130097                 1.0                   939                     6   

        FAMI_EDUCACIONMADRE  ESTU_DEDICACIONLECTURADIARIA  
0                       2.0                           0.0  
1                       6.0                           0.0  
2                       3.0                           1.0  
3                       3.0                           0.0  
4                       1.0                           3.0  
...                     ...                           ...  
130093                  5.0                           4.0  
130094                  4.0                           2.0  
130095                  4.0                           0.0  
130096                  1.0                           3.0  
130097                  8.0                           1.0  

[130098 rows x 13 columns]

### WE ARE READY TO APPLY A MACHINE LEARNING MODEL

In [28]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X_train_final, y_train)


LinearRegression()

In [29]:
y_pred = model.predict(X_valid_final)

y_pred

array([0.4543575 , 0.52459598, 0.50195956, ..., 0.54060531, 0.466483  ,
       0.54126632])

#### Evaluating the model

In [30]:
from sklearn.metrics import mean_absolute_error,r2_score,mean_squared_error

ma=mean_absolute_error(y_valid,y_pred)
r2=r2_score(y_valid,y_pred)
ms=mean_squared_error(y_valid,y_pred)

print(ma)

print(r2)


print(ms)

0.07104476948248735
0.23102178819073005
0.007679505382785207


In [37]:
print(model.feature_names_in_)
print(model.coef_) # weight of each variable in the prediction

['ESTU_GENERO_F' 'ESTU_GENERO_M' 'FAMI_TIENEINTERNET_No'
 'FAMI_TIENEINTERNET_Si' 'FAMI_TIENECOMPUTADOR_No'
 'FAMI_TIENECOMPUTADOR_Si' 'EDAD' 'FAMI_ESTRATOVIVIENDA'
 'FAMI_PERSONASHOGAR' 'COLE_MCPIO_UBICACION' 'COLE_DEPTO_UBICACION'
 'FAMI_EDUCACIONMADRE' 'ESTU_DEDICACIONLECTURADIARIA']
[ 4.21088738e+08  4.21088738e+08  2.98936657e+08  2.98936657e+08
  2.25982078e+05  2.25982097e+05 -2.10822411e-02  4.01131995e-03
 -7.71096349e-03 -1.22189522e-05  1.85010955e-04  9.37865302e-03
  1.44631118e-02]


#### Conclusions


-Before the evalution of the model I concluded data isn´t big enough to get more accurate predictors.
-An R^2 value of 0.2310226251792521 suggests that the model explains approximately 23.1% of the variability present in the data, which is not a very good fit.


#### Top Fields
Fields ranked by their contribution to the prediction results: using model.coef_ and model.feature_names_in_ and some transformations we get for each field:
porcentajes = [
    14.3, 13.2, 13.1, 11.0, 10.5, 8.0, 8.0, 7.6, 7.4, 6.8
]

In [36]:
import plotly.graph_objects as go

# Datos de la tabla
categorias = [
    'COLE_DEPTO_UBICACION', 'COLE_MCPIO_UBICACION', 'FAMI_EDUCACIONMADRE',
    'ESTU_DEDICACIONLECTURADIARIA', 'EDAD', 'ESTU_GENERO',
    'FAMI_ESTRATOVIVIENDA', 'FAMI_TIENEINTERNET', 'FAMI_TIENECOMPUTADOR', 'FAMI_PERSONASHOGAR'
]

porcentajes = [
    14.3, 13.2, 13.1, 11.0, 10.5, 8.0, 8.0, 7.6, 7.4, 6.8
]

# Crear el gráfico de barras horizontales
fig = go.Figure()

fig.add_trace(go.Bar(
    y=categorias,
    x=porcentajes,
    marker_color='blue',
    opacity=0.6,
    orientation='h'
))

# Personalización del gráfico
fig.update_layout(
    title='Fields ranked by their contribution to the prediction results',
    yaxis_title='Fields',
    xaxis_title='%',
    width=800,  # Ajusta el ancho del gráfico (en píxeles)
    height=600,  # Ajusta el alto del gráfico (en píxeles)
    margin=dict(l=100, r=50, t=60, b=50)  # Ajusta los márgenes del gráfico para evitar recortes de etiquetas
)

# Mostrar el gráfico
fig.show()
